usermount/usermount.man
.TH usermount 1 "February 1996"
.SH NAME
usermount, userumount \- Filesystem mounting and unmounting utilities

.SH SYNOPSIS
.B usermount
\fIdevice\fP
.br
.B usermount
\fIdirectory\fP
.br
.B usermount -version
.PP
.B userumount
\fIdevice\fP
.br
.B userumount
\fIdirectory\fP
.br
.B userumount -version
.sp
.SH DESCRIPTION
Usermount and userumount enable ordinary users to mount and unmount certain
filesystems. Either the special \fIdevice\fP to be (un)mounted or the corresponding
mount point \fIdirectory\fP have to be given as arguments to usermount and
userumount.
The file \fI/usr/local/lib/userfstab\fP determines, which devices users
may mount on which mountpoints. The format of the \fIuserfstab\fP file is the same
as that of \fI/etc/fstab\fP (see the manpage \fIfstab\fP(4)). A typical \fIuserfstab\fP
file might look as follows:
.sp
.nf
	/dev/fd0		/floppy	msdos	noexec	0	0
	/dev/cdrom	/cdrom	iso9660	ro	0	0
.fi
.sp
In this example users are allowed to mount the floppy drive \fI/dev/fd0\fP (with DOS
formatted floppies) on the directory \fI/floppy\fP, and to mount the CD-ROM read-only
on \fI/cdrom\fP. The floppy is mounted with the command \fIusermount /floppy\fP or
\fIusermount /dev/fd0\fP.
.PP
When usermount and userumount have been successfull (un)mounting a device, the
\fI/etc/mtab\fP file is updated accordingly. When a device cannot be mounted
\fIread-write\fP, usermount will try to mount it \fIread-only\fP.
.PP
If usermount and userumount have been compiled with the \fI-DLOGFILE=\\"<log-file>\\"\fP
option, all (un)mounting actions are logged in <log-file>. To find out if this is the
case, run \fIusermount -version\fP.
.PP
In principle, users may only unmount those devices with userumount which they
themselves mounted in the first place. This does not apply when userumount cannot
determine who mounted the device by looking at the ownership of the mount point,
e.g. for non-\fImsdos\fP and non-\fIhpfs\fP filesystems mounted read-only.
.PP
.B Caution:
If usermount has been compiled with the \fI-DSET_OWNER\fP option, the user is made
the owner of filesystems he is mounting read-write, in order to make sure that he has
write permissions. To find out if this is the case, run \fIusermount -version\fP.
.PP
.B
Caution:
Usermount and userumount have to be \fIsetuid root\fP in order to work. It is important
that the \fIuserfstab\fP file is owned by root and not writable for any other user. You
can restrict the usage of user(u)mount to a certain group of users by making user(u)mount
non-executable for others than members of this group. 
.SH "FILES"
.TP 4
.B /usr/local/lib/userfstab
Contains table of devices mountable with usermount.
.TP 4
.B /etc/mtab
Contains table of devices currently mounted with \fImount\fP or usermount.
.SH "SEE ALSO"
\fImount\fP(8), \fIfstab\fP(4), \fImoxfm(1)\fP
.SH "CAVEATS AND BUGS"
Usermount does not support auto filesystem detection. Userumount does not
work properly for \fImsdos\fP and \fIhpfs\fP filesystems mounted with the
\fIuid=\fP option. If you find more bugs, email mai@sugra.desy.de.
.SH "COPYRIGHT"
Copyright (c) 1995, 1996 by Oliver Mai
.SH "AUTHOR"
Usermount and userumount were written by Oliver Mai (mai@sugra.desy.de) and are
designed be used with the \fImoxfm\fP filemanager.


ChangeLog
2001-09-03  root  <Dr.Graef@t-online.de>

	* changed program info (version number, www pointer)

	* fixed a bug in FmFw.c (eliminate trailing newline from ctime
	return value)

	* moxfm config files now go to .moxfm

	* face-lifting using modern apps and 3dicons, also tweaked some of
	the resources

	* added magic.c patches (from xplore/xfm)

	* added Bob Woodside's 1.0.1 patches (see below)

2001-08-15:	Version 1.0.1:

		New features:
                None.

                Changes:
                This is a minimal maintenance release designed to correct a
                few problems that have cropped up since Oliver left off work
                on moxfm, and produce a viable current distribution of the
                source.

                Bug Fixes:
                - regexp/regexp.c: Commented out argument-less declaration for
                strchr(), which causes compile failures with recent compilers.
                I know of no systems on which this declaration is needed any
                longer.
                - regexp/regsub.c: Commented out argument-less declaration for
                strncpy().
                - src/Common.h: Added logic to force define for SYSV on recent
                Linux systems, so that getwd will be properly redefined as
                getcwd. getwd is a dangerous function that generally should not
                be used, and recent versions of gcc issue a warning to this
                effect.
                - src/FmFw.c, src/FmIcons.c: Added a workaround for a failure
                of double-clicks to work under certain window managers/window
                manager modes, due to ill-synchronized passive grabs by the
                window manager on a parent window. This "fix" is based on a
                similar one made by Till Straumann for xfm.

**************
OLD ChangeLog:
**************

Date (yy-mm-dd):

95-08-17:	Moxfm v. 0.9 is available for public
		use for the first time via WWW. It is
		based on xfm v. 1.3.2.

95-08-21:	Version 0.9pl1: Bugfixes concerning 
		the state of the mount table window at
		startup 

95-08-22:	Version 0.91: Changes in function
		fnmultimatch() to allow for escaping
		blanks.

95-08-24:	Version 0.92: Changes in file window
		update code.

95-08-25:	Version 0.93: Code for updating file
		windows changed again. Mount point
		detection improved.

95-08-26:	Version 0.94: When copying, moving,
		linking or deleting fails for a number
		of files, the screen is no longer flooded
		with error message dialogs, but the
		user has the chance to cancel the
		operation. Changes in the code for
		recursive copying. Bugfixes.

95-08-28:	Version 0.94pl1: Bugfix concerning
		status display in connection with file
		selections.

95-08-29:	Version 0.94pl2: Bugfix in filesystem
		info dialog.

95-08-30:	Version 0.95: Changes in FmDragDrop.c
		in order to make moxfm compatible
		with future versions which (hopefully)
		will support copying files to/from a
		remote machine. Bugfix concerning
		-dir command line argument.

95-08-31:	Version 0.95pl1: Fixes in device mounting
		in FmFwActions.c and FmOps.c.

95-09-02:	Version 0.96: Major code clean-up of
		FmFwActions.c, also affecting other
		modules. Bugfixes here and in FmOps.c.

95-09-08:	Version 0.97: FIRST VERSION UPLOADED TO
		FTP-SERVER.
		
		- You cannot drop a file onto itself
		  any longer.
		
		Bugfixes:
		- When a filename is specified for an
		  application button, the filename is
		  passed as first argument also for
		  drop actions. This implies e.g. that
		  the drop action for the trashcan should
		  be
		  "shift;mv -f -b -V numbered $* ~/.trash"
		  now (as for xfm).
		- The icon names are set correctly now.

95-09-09:	Bugfix in usermount utility: userfstab
		file is checked for all mount options,
		not only "read-only" now.

95-09-18:	Version 0.98:

		New features:
		- There is a popup menu of the directory
		  name text field now (thanks to Steve
		  Blake for idea.)
		- The "File" pulldown menu has a "Perform
		  action" item now.
		- The mount table window has a "Close"
		  button.
		- Links are always made to *relative*
		  paths.
		- New resources saveOnExit and newWinOnDirPopup.
		
		Bugfixes:
		- Moxfm should hopefully compile on OSF/1
		  and SVR4 systems now (thanks to Albert
		  Graef and Volker Mertens).
		- Moxfm compiles with Motif 2.0 now.
		- Moxfm doesn't crash any longer when it
		  cannot allocate enough colors for the
		  pixmaps (thanks to Detlef Schmier).
		- The icon names should really be correct
		  now.
		  
		Changes:
		- When the autoSave feature of the
		  application windows is enabled, the
		  configuration is saved directly after
		  every change and not first when moxfm
		  exits.
		- The *relative* paths of dropped files
		  are passed to application buttons with
		  specified directories now.
		- There is a "Perform action" item in the
		  pulldown menu of directory icons, too.
		- New suppressWarnings resource instead of
		  suppressXtWarnings resource.
		- The names of the application group
		  configuration files don't start with "xfm-"
		  any longer.

96-05-16:	Version 1.0:

		New features:
		- Files and applications can be placed as icons
		on the root window.
		- Added automatic application window updates along
		with a checkApplicationFiles resource, and an applBoxIcon
		resource for specifying the standard application group
		icon. Furthermore, newly created application groups now
		point back to the group they were created in.
		(ag@muwiinfa.geschichte.uni-mainz.de)
		- Added a facility for monitoring files during
		automatic updates, along with the new xfmmon
		configuration file (AG).
		- You can drop user defined objects (listed in
		targets configuration file) on application windows.
		- You can switch off magic headers for certain
		devices and file windows.
		- You can move files across devices now.
		- The file type configuration file can be reread
		during runtime.

usermount/usermount.c
#include <stdio.h>
#include <unistd.h>
#include <errno.h>
#include <mntent.h>
#include <sys/mount.h>
#include <linux/fs.h>

#ifdef LOGFILE
#include <time.h>
#include <pwd.h>
#include <sys/types.h>
#endif

#define VERSION "1.0"
#define UFSTAB "/usr/local/lib/userfstab"
#define MTAB "/etc/mtab"

#ifndef MS_SYNCHRONOUS
#define MS_SYNCHRONOUS MS_SYNC
#endif

#ifdef LOGFILE
void logAction(char *device, int err);

void logAction(char *device, int err)
{
 struct passwd *pwent;
 FILE *flog = fopen(LOGFILE, "a");
 time_t now;

 if (flog)
 {
     pwent = getpwuid(getuid());
     now = time(NULL);
     fprintf(flog, "%s", asctime(localtime(&now)));
     if (err)
	 fprintf(flog, "Unsuccessfull trial to mount %s by %s: %s\n", device, pwent->pw_name, sys_errlist[err]);
     else
	 fprintf(flog, "%s mounted by %s\n", device, pwent->pw_name);
     fclose(flog);
 }
 else
     fprintf(stderr, "Unable to open log file %s for writing.\n", LOGFILE);

}
#endif

int main(int argc, char *argv[])
{
 struct mntent *entry;
 FILE *ufstab;
 FILE *mtab;
 unsigned long rwflag = MS_MGC_VAL;
 int found = 0, errfl = 0, readonly = 0;

 if (argc != 2)
 {
     fprintf(stderr, "usermount - Usage: usermount <device>\n");
     exit(-2);
 }
 if (!strcmp(argv[1], "-version"))
 {
     printf("usermount version %s\nUser filesystem table: %s\n", VERSION, UFSTAB);
#ifdef SET_OWNER
     printf("Caution: The user is made the owner of read-write mounted filesystems.\n");
#else
     printf("The ownerships of mounted filesystems other than DOS FAT and HPFS are not changed.\n");
#endif
#ifdef LOGFILE
     printf("(Un)mount actions are logged in %s.\n", LOGFILE);
#else
     printf("(Un)mount actions are not logged.\n");
#endif
     exit(0);
 }

 if (!(ufstab = setmntent(UFSTAB, "r")))
 {
     fprintf(stderr, "usermount: Could not open %s\n", UFSTAB);
     exit(-1);
 }
 while ((entry = getmntent(ufstab)))
 {
     if (!strcmp(argv[1], entry->mnt_fsname) || !strcmp(argv[1], entry->mnt_dir))
     {
	 found = 1;
	 break;
     }
 }
 if (!found)
 {
     fprintf(stderr, "usermount: no entry matching %s found in %s\n", argv[1], UFSTAB);
     endmntent(ufstab);
     exit(-3);
 }
 if (hasmntopt(entry, "ro"))
 {
     readonly = 1;
     rwflag |= MS_RDONLY;
 }
 if (hasmntopt(entry, "nosuid"))
     rwflag |= MS_NOSUID;
 if (hasmntopt(entry, "nodev"))
     rwflag |= MS_NODEV;
 if (hasmntopt(entry, "noexec"))
     rwflag |= MS_NOEXEC;
 if (hasmntopt(entry, "sync"))
     rwflag |= MS_SYNCHRONOUS;
 if (mount(entry->mnt_fsname, entry->mnt_dir, entry->mnt_type, rwflag, NULL))
 {
     if (errno == EROFS || errno == EACCES)
     {
	 readonly = 1;
	 rwflag |= MS_RDONLY;
	 if (mount(entry->mnt_fsname, entry->mnt_dir, entry->mnt_type, rwflag, NULL))
	     errfl = 1;
     }
     else errfl = 1;
     if (errfl)
     {
	 fprintf(stderr, "usermount: Could not mount %s on %s: %s\n", entry->mnt_fsname, entry->mnt_dir, sys_errlist[errno]);
#ifdef LOGFILE
	 logAction(entry->mnt_fsname, errno);
#endif
	 endmntent(ufstab);
	 exit(errno);
     }
 }
 if  (!(mtab = setmntent(MTAB, "r+")) || addmntent(mtab, entry))
 {
     fprintf(stderr, "usermount: Warning - error updating %s: %s\n", MTAB, sys_errlist[errno]);
     endmntent(mtab);
 }
#ifdef SET_OWNER
 if (!readonly && strcmp(entry->mnt_type, MNTTYPE_MSDOS) && strcmp(entry->mnt_type, MNTTYPE_HPFS))
     if (chown(entry->mnt_dir, getuid(), getgid()))
	 fprintf(stderr, "usermount: Warning - unable to set ownership of %s\n", entry->mnt_dir);
#endif

#ifdef LOGFILE
	 logAction(entry->mnt_fsname, 0);
#endif

 endmntent(ufstab);

 return 0;
}
